arelle/XPathParser.py

```diff
@@ -1,227 +1,343 @@
 '''
 See COPYRIGHT.md for copyright information.
 '''
+from __future__ import annotations
+
 import sys
-from numbers import Number
+import time
+import traceback
+from collections.abc import Callable, Iterable, Sequence
+from decimal import Decimal
+from typing import Any, TYPE_CHECKING, Union
+from xml.dom import minidom
 
-from arelle import PythonUtil # define 2.x or 3.x string types (only needed when running as unit test from __main__
-from arelle.PluginManager import pluginClassMethods
 from pyparsing import (
-    Word, Keyword, alphas, ParseException, ParseSyntaxException,Literal, CaselessLiteral,Combine, Optional, nums, Forward, Group, ZeroOrMore, StringEnd, alphanums,ParserElement, quotedString,
-    delimitedList, Suppress, Regex
+    CaselessLiteral,
+    Combine,
+    Forward,
+    Group,
+    Keyword,
+    Literal,
+    Opt,
+    ParseBaseException,
+    ParseException,
+    ParseResults,
+    ParseSyntaxException,
+    ParserElement,
+    Regex,
+    StringEnd,
+    Suppress,
+    Word,
+    ZeroOrMore,
+    alphanums,
+    alphas,
+    delimited_list,
+    nums,
+    quoted_string,
 )
+
+from arelle import ModelValue, XbrlConst, XmlUtil
 from arelle.Locale import format_string
-import time, xml.dom, traceback
-from decimal import Decimal
-from arelle import (XmlUtil, ModelValue, XbrlConst)
-FunctionIxt = None
+from arelle.PluginManager import pluginClassMethods
+
+if TYPE_CHECKING:
+    from arelle.ModelFormulaObject import ModelFormulaResource
+    from arelle.ModelXbrl import ModelXbrl
+    from arelle.ModelManager import ModelManager
+    from arelle.ModelObject import ModelObject
+    from arelle.ModelValue import QName
+    from arelle.XPathContext import XPathContext, XPathException
+    from arelle.typing import TypeGetText
+
+    _: TypeGetText  # Handle gettext
+
+    FormulaToken = Union[
```

Comment:
This feels really helpful, what do you think about breaking out into `FormulaGrammar.py` or something that makes it obvious that its defining just the tokens? As a side note, with this being so compilery I don't know if you are already planning some of this but it seems easier to understand for people if we eventualize organize this more like a traditional parser with the  cutoffs, `grammar -> tokenizer -> ast ... ect`

Reply:
That'll be the first commit in #536